### ISSUE TITLE
Future work: NN to train rank learner  under total utility loss

### ISSUE THREAD

**Issue by christianadriano** [hpi-sam/robust-marl4sas#47]
Have a network for the rank learner to decide which policy the agent should use to determine the component to be fixed. Assuming that each agent has learned a policy by training a network (currently and actor-critic architecture), the rank learner can itself learn whether a given agent policy is performing well or not. To know that, the rank learner could use the signal from mRubis, e.g., loss in total utility, and map to the corresponding underperforming fixes. A fix might have underperformed for two reasons: the probability of success for that fix was too way off from the actual or the predicted utility of the corresponding component was too far from the actual one. The latter information is promptly available from mRubis, whereas the former not. The rank learner would need to build a loss function that can distinguish whether the problem is in the utility prediction or in the fix success probability prediction. 

In case that the Rank learner determines that a given agent has an underperforming policy, the Rank learner can recommend a set of measures: look for a better policy among the other agents, transfer a better policy to the current underperforming agent, or trigger a retraining of the current agent's policy.

How the Rank learner would learn? Possible inputs: Each agents probability distribution for the current state of this shop. Possible Output: Softmax output over all possible networks. The signal (loss function) could be based on the total utility loss across mRubis shops.

**Comment by jokrah** [comment]
I agree with the idea of having a NN which decides what to do next: Do nothing (the agent is doing a good job), fix the A2C (the agent is not returning the correct fix), fix the ranking (RidgeRegression has to be retrained).
To sum up your input:
Inputs: actor_value, critic_value, loss, reward
Output: [NoOp, Fix A2C, Fix Ranking]

A possible approach for "Fix A2C":
This could also be seen as a slightly different approach: We could have a NN that is acting as a layer to choose the best suitable NN for the next step in order to prevent a wrong action. For this, we could see the agents as offering their trained NN and the rank learner is deciding which network is best suited for a specific shop based on the last few steps. The same applies for another NN where the RankLearner has different RidgeRegression versions to choose from and decides for a shop which to use based on the last steps. 
But the longer I think about the "Decide-NN" approach, I think it is not well suited. 

However, learning when to learn seems reasonable.
